akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala

```diff
@@ -186,7 +186,10 @@ import scala.util.control.NonFatal
       override def offer(element: T): Future[QueueOfferResult] = {
         val p = Promise[QueueOfferResult]
         callback.invokeWithFeedback(Offer(element, p))
-          .onFailure { case NonFatal(e) ⇒ p.tryFailure(e) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+          .failed.foreach {
+            case NonFatal(e) ⇒ p.tryFailure(e)
+            case _           ⇒ ()
+          }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
```

Comment:
This allocates an extra lambda per call/element (inside `.failed`), replace with: 

```
          .onComplete {
            case scala.util.Success(_)           ⇒
            case scala.util.Failure(NonFatal(e)) ⇒ p.tryFailure(e)
          }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
```

to get performance on par with master, same for the sink (and potential other hot spots where we are using `.failed` now

Reply:
Difference for invoke with feedback benchmark is 50k ops/s failed.foreach compared to 80k ops/s with onComplete across both 2.12 and 2.13

Reply:
Don't forget non fatal failures.